crates/polars-ops/src/series/ops/index_of.rs

```diff
@@ -0,0 +1,137 @@
+}
+
+/// Find index where predicate is true.
+fn index_of_predicate<P, T>(ca: &ChunkedArray<T>, predicate: P) -> Option<usize>
+where
+    T: PolarsNumericType,
+    P: Fn(&T::Native) -> bool,
+{
+    let mut index = 0;
+    for chunk in ca.chunks() {
+        let chunk = chunk
+            .as_any()
+            .downcast_ref::<PrimitiveArray<T::Native>>()
+            .unwrap();
+        if chunk.validity().is_some() {
+            for maybe_value in chunk.iter() {
+                if maybe_value.map(&predicate) == Some(true) {
+                    return Some(index);
+                } else {
+                    index += 1;
+                }
+            }
+        } else {
+            // No nulls, so we can simplify:
+            for value in chunk.values_iter() {
+                if predicate(value) {
+                    return Some(index);
+                } else {
+                    index += 1;
+                }
+            }
+        }
+    }
+    None
+}
+
+impl<'a, T> ChunkSearch<'a, T::Native> for ChunkedArray<T>
+where
+    T: PolarsNumericType,
+{
+    fn index_of(&'a self, value: Option<T::Native>) -> Option<usize> {
+        // A NaN is never equal to anything, including itself. But we still want
+        // to be able to search for NaNs, so we handle them specially.
+        if value.map(|v| v.is_nan()) == Some(true) {
+            return index_of_predicate(self, |v| v.is_nan());
+        }
+
+        // Searching for an actual value:
+        if let Some(value) = value {
+            return index_of_predicate(self, |v| *v == value);
+        }
+
+        // Searching for null:
+        let mut index = 0;
+        for chunk in self.chunks() {
+            let length = chunk.len();
+            if let Some(bitmap) = chunk.validity() {
+                let leading_ones = bitmap.leading_ones();
+                if leading_ones < length {
+                    return Some(index + leading_ones);
+                }
+            } else {
+                index += length;
+            }
+        }
+        None
+    }
+}
+
+/// Try casting the value to the correct type, then call index_of().
+macro_rules! try_index_of_numeric_ca {
+    ($ca:expr, $value:expr) => {{
+        let ca = $ca;
+        let value = $value;
+        if value == AnyValue::Null {
+            return Ok(ca.index_of(None));
+        }
+        let cast_value = cast_if_lossless(&value, ca.dtype());
+        if cast_value.is_some() {
+            ca.index_of(cast_value.map(|v| v.extract().unwrap()))
+        } else {
+            // We can't cast the searched-for value to a valid data point within
+            // the dtype of the Series we're searching, which means we will
+            // never find that value.
+            None
+        }
+    }};
+}
+
+/// Cast to the dtype, but return ``None`` if the casting changed the value.
+pub fn cast_if_lossless<'a>(value: &'a AnyValue, dtype: &'a DataType) -> Option<AnyValue<'a>> {
```

Comment:
We should already have casted in IR coversion. In puts should be cast to supertype, then we can have that as an invariant here.

Reply:
Ah, so, I did casting manually on purpose: my presumption was if the code casts in the IR, it also casts the Series, which seems expensive, whereas here we only have to cast the value we're searching for. Does that make sense? If so I'll document, if not I'll switch.

Reply:
We could cast only the search value, in the IR.

Reply:
OK. Where is the relevant code path for that?